### DISCUSSION TITLE
Angular Query: How to handle persistence / restoration race condition

### DISCUSSION THREAD

**Discussion by OmerGronich** [TanStack/query#7570] [Q&A]
I'm trying to figure out how to use angular query with indexedDB persistence. I'm running into a race condition issue that I'm not sure exactly how to solve. 

Just to explain my use case and what I'm trying to do exactly, I have requests that run with [APP_INITIALIZER](https://angular.dev/api/core/APP_INITIALIZER?tab=usage-notes),  requests for metadata or configuration data that only need to run once and be available across the app ASAP.
I'm trying to cache these requests using angular query so they don't delay the user if they try to open multiple tabs. Because these requests might return very large objects I'm using indexedDB as the persister. 

[issue repro](https://stackblitz.com/edit/tanstack-query-hztusg?file=src%2Fapp%2Fapp.config.ts) 
In my attempts to make this work I encountered an issue where the cache was being ignored, you can see this in action in the stackblitz above, if you open the chrome console you will see the logs where the query fn is called before the `persistQueryClientRestore` restores the data from indexedDB.
![image](https://github.com/TanStack/query/assets/59396491/ef6d758c-5b0a-4c02-bfc7-f5b9aa8e4821)

After searching for similar discussions I encountered this [comment](https://github.com/TanStack/query/discussions/5867#discussioncomment-6731634) by @TkDodo that explains why this scenario occurs: `ensureQueryData` does not have a built-in mechanism that tracks the restoring state. 

I did find a hacky solution, to bootstrap the application only after the restoring process is finished: 
```
persistQueryClient({ queryClient, persister });
persistQueryClientRestore({
  queryClient,
  persister,
}).finally(() => {
  bootstrapApplication(BasicExampleComponent, appConfig).catch((err) =>
    console.error(err)
  );
});
```
[possible solution repro](https://stackblitz.com/edit/tanstack-query-qfmurj?file=src%2Fmain.ts)

This does work, but feels very  wrong. 

Is there a better way to go about this?

**Comment by TkDodo** [comment]
rendering the application after restore has completed seems fine to me. In React, we have a dedicated `<PersistQueryClientProvider>` that puts the app into `restoring` mode so that no fetches are being made. I guess the angular adapter would need something similar. @arnoud-dv FYI

**Reply by arnoud-dv** [reply]
I haven't tried any persistence myself but I've read, [for example here](https://github.com/TanStack/query/discussions/6293#discussioncomment-9230475) that others used `@tanstack/query-sync-storage-persister` [and this framework agnostic package](https://tanstack.com/query/latest/docs/framework/react/plugins/createPersister) with good results. That might be either a solution or a good starting point to make your own implementation.

Also if anyone wants to contribute persistence to the Angular adapter I would welcome a PR.

**Reply by rbalet** [reply]
Hey @arnoud-dv .

I would need the persistQuery for my own app, and I'm motivated to do a PR. (If you're still open to that)
The former `@tanstack/query-sync-storage-persister` is now being deprecated in favour of [@tanstack/query-async-storage-persister](https://tanstack.com/query/v4/docs/framework/react/plugins/createAsyncStoragePersister).

How do you think I should tackle this problem ?
